components/root/index.js

```diff
@@ -23,17 +26,44 @@ import {isMobile} from 'utils/utils.jsx';
 import Root from './root.jsx';
 
 function mapStateToProps(state) {
+    const getCategory = makeGetCategory();
     const config = getConfig(state);
     const showTermsOfService = shouldShowTermsOfService(state);
     const plugins = state.plugins.components.CustomRouteComponent;
     const products = state.plugins.components.Product;
+    const userId = getCurrentUserId(state);
 
-    const teamId = LocalStorageStore.getPreviousTeamId(getCurrentUserId(state));
+    const teamId = LocalStorageStore.getPreviousTeamId(userId);
     const permalinkRedirectTeam = getTeam(state, teamId);
-    const taskListStatus = getBool(state, OnboardingTaskCategory, OnboardingTaskList.ONBOARDING_TASK_LIST_SHOW);
-    const isUserFirstAdmin = isFirstAdmin(state);
-    const isMobileView = isMobile();
-    const showTaskList = isUserFirstAdmin && taskListStatus && !isMobileView;
+
+    let showTaskList = false;
+    let firstTimeOnboarding = false;
+
+    // validation to avoid execute logic on first load which has no preferences values on global store
+    if (Object.keys(state.entities.preferences.myPreferences).length > 0) {
+        // conditions to validate scenario where users (initially first_admins) had already set any of the onboarding task list preferences values.
+        // We check wether the preference value exists meaning the onboarding tasks list already started no matter what the state of the process is
+        const onboardingPreferences = getCategory(state, OnboardingTaskCategory);
+        const hasUserStartedOnboardingTaskListProcess = onboardingPreferences?.some((pref) =>
+            pref.name === OnboardingTaskList.ONBOARDING_TASK_LIST_SHOW || pref.name === OnboardingTaskList.ONBOARDING_TASK_LIST_OPEN);
+
+        // This condition verifies existing users hasn't finished nor skipped legacy next steps or there are still steps not completed
+        const legacyStepsPreferences = getCategory(state, Preferences.RECOMMENDED_NEXT_STEPS);
+        const hasLegacyNextStepsPrefs = legacyStepsPreferences?.some((pref) =>
+            pref.name === RecommendedNextSteps.SKIP || pref.name === RecommendedNextSteps.HIDE);
+
+        const existingUserHasntFinishedNorSkippedLegacyNextSteps = hasLegacyNextStepsPrefs && showOnboardingTaskListToExistingUsers(state);
+
+        const completelyNewUserForOnboarding = !hasUserStartedOnboardingTaskListProcess && !hasLegacyNextStepsPrefs;
+
+        if (completelyNewUserForOnboarding || existingUserHasntFinishedNorSkippedLegacyNextSteps) {
+            firstTimeOnboarding = true;
+        }
+
+        const taskListStatus = getBool(state, OnboardingTaskCategory, OnboardingTaskList.ONBOARDING_TASK_LIST_SHOW);
+        const isMobileView = isMobile();
+        showTaskList = (firstTimeOnboarding || taskListStatus) && !isMobileView;
```

Comment:
Do we have any test coverage for all these cases? We may have an opportunity to extract this as a function and add some unit test - this is a vital code path that can cause every existing user to get the checklist on upgrade or no one getting the checklist too :D

Reply:
correct! this is the piece of code that handles all the logic to determine when to show/hide the task list. I will extract it to a function and write unit test for it.